### PR TITLE
Add lazy::Shape::numel()

### PR DESCRIPTION
**Pull request by wconstab**
Summary: Add a convenience to lazy::Shape for counting the number of elements (by multiplying out the dimensions).  This is a method on Tensor, and in switching other lazy tensor shape utils to use aten shape inference, we need numel counts.

Test Plan: add unit tests

Differential Revision: D32409138

### CODE REVIEW

File: torch/csrc/lazy/core/shape.cpp [hunk 1]
```diff
@@ -33,5 +33,15 @@ std::vector<Shape> convertShapes(
   return shape;
 }
 
+size_t Shape::numel() const {
+  size_t elts = dim() > 0 ? 1 : 0;
+  for (size_t i = 0; i < dim(); i++) {
```
Comment (desertfire) [inline torch/csrc/lazy/core/shape.cpp:38]: Nit: access sizes_ directly with a range-based loop?

**Review by desertfire — APPROVED** [review]

File: torch/csrc/lazy/core/shape.cpp [hunk 2]
```diff
@@ -33,5 +33,15 @@ std::vector<Shape> convertShapes(
...
+    // TODO(whc) is this check both necessary, and correct?
+    TORCH_CHECK(size(i) > 0);
```
Comment (alanwaketan) [inline torch/csrc/lazy/core/shape.cpp:40]: Maybe `TORCH_INTERNAL_ASSERT` if you are not sure?

File: torch/csrc/lazy/core/shape.cpp [hunk 3]
```diff
@@ -33,5 +33,15 @@ std::vector<Shape> convertShapes(
   return shape;
 }
 
+size_t Shape::numel() const {
```
Comment (alanwaketan) [inline torch/csrc/lazy/core/shape.cpp:36]: Wondering if we want to cache the result somewhere? Also, should we consider integer overflow like: https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L2207?

Comment (wconstab) [reply torch/csrc/lazy/core/shape.cpp:36]: Caching could be a good idea, I'm not sure if it's called often enough to warrant the complexity tradeoff, but it seems simple enough given Shape class is small for the time being.

Integer overflow probably isn't a likely scenario for our tensors.  Note the comment they have for the compute_safe_numel version:
`Use only for tensors with sparse layouts because only
   * sparse tensor are likely to have sizes that may lead to integer
   * overflow when computing numel`
   
We don't support sparse-lazy tensors.

File: torch/csrc/lazy/core/shape.cpp (hunk 2 above)
Comment (wconstab) [reply torch/csrc/lazy/core/shape.cpp:40]: I'm confused- Isn't TORCH_INTERNAL_ASSERT a stronger statement of a case that is absolutely a bug if false?  And TORCH_CHECK is somewhat more relaxed?  In any case, i think they do the same thing, but, I am not that sure when to use one or the other.  

Regardless, I'm not sure this check should be here at all.  IT appears that size==0 is actually valid in pytorch, and numel() should return zero in those cases. So i think I need to delete this. 

```
>>> x = torch.ones((1,0,2))
>>> x.numel()
0
```

Comment (alanwaketan) [reply torch/csrc/lazy/core/shape.cpp:40]: I think the difference is that `TORCH_CHECK` is user facing and `TORCH_INTERNAL_ASSERT` is debug assert. I'm not sure about which one is stronger since they all just crash the program. For error logging, you might want to use `LOG(ERROR)`.

File: torch/csrc/lazy/core/shape.cpp (hunk 3 above)
Comment (alanwaketan) [reply torch/csrc/lazy/core/shape.cpp:36]: I have a use case in `LTCTensorImpl` that will set it's `numel` by using this `numel`. So I think caching would be a good idea.

File: torch/csrc/lazy/core/shape.cpp [hunk 4]
```diff
@@ -33,5 +33,13 @@ std::vector<Shape> convertShapes(
...
+  size_t elts = dim() > 0 ? 1 : 0;
```
Comment (alanwaketan) [inline torch/csrc/lazy/core/shape.cpp:37]: I'm not sure setting `elts` to `0` is actually necessary given https://github.com/pytorch/pytorch/blob/master/c10/core/TensorImpl.h#L2195 doesn't do that.

Comment (alanwaketan) [reply torch/csrc/lazy/core/shape.cpp:37]: Probably safer to keep it given TensorImpl.h might have other special handlings.

Comment (wconstab) [reply torch/csrc/lazy/core/shape.cpp:37]: i'm not sure, but at least my way produces the behavior that's expected (tensor of size (1,0,2) reports numel()==0)
@bdhirsh any idea if i should do this differently?

File: torch/csrc/lazy/core/shape.cpp (hunk 3 above)
Comment (wconstab) [reply torch/csrc/lazy/core/shape.cpp:36]: hmm.  now you make me wonder (a) still whether caching is actually important for perf (should we measure first?). (b) more importantly, should it be at the Shape layer or at the TensorImpl layer?  I wonder if TensorImpl layer has more visibility into whether there is an actual at::Tensor data or an IR/computation backing the lazy tensor, and caching in LTCTensorImpl would allow taking advantage of that knowledge.

Comment (alanwaketan) [reply torch/csrc/lazy/core/shape.cpp:36]: `LTCTensorImpl` caches this value by default, it only queries shape when the generation increments. So I guess it's okay for `LTCTensorImpl`. It's true that it's just my gut feeling that caching could be benefit.

**Review by alanwaketan — APPROVED** [review]
